fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/PredicatedTransformer.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.biome;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import it.unimi.dsi.fastutil.doubles.DoubleArrayList;
+import it.unimi.dsi.fastutil.doubles.DoubleList;
+
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.biome.layer.util.LayerRandomnessSource;
+
+import net.fabricmc.fabric.api.biomes.v1.BiomeNeighborsPredicate;
+
+final class PredicatedTransformer {
+	private final List<PredicatedBiomeEntry> predicates = new ArrayList<>();
+
+	void addPredicatedBiome(Biome biome, BiomeNeighborsPredicate predicate, double chance) {
+		predicates.add(new PredicatedBiomeEntry(biome, predicate, chance));
+	}
+
+	Biome transform(Biome biome, LayerRandomnessSource random, Collection<Biome> borders) {
+		if (predicates.isEmpty()) {
+			return biome;
+		} else if (predicates.size() == 1) {
+			PredicatedBiomeEntry predicate = predicates.get(0);
+
+			if (predicate.test(borders, random)) {
+				return predicate.getBiome();
+			}
+
+			return biome;
+		}
+
+		List<PredicatedBiomeEntry> truePredicates = new ArrayList<>();
```

Comment:
We could just collect the biomes here in `List<Biome> candidates` instead. I'm not sure if it's worth doing that though.

Reply:
the true predicate carries a weight information. The biome itself doesn't.

Reply:
But we don't access it after collecting the list.

Reply:
The weight is a double, not an int, so you cannot use an integer-indexed list easily.

Reply:
But that's what the `DoubleList` is for. We manage a parallel list of the predicates and one of the doubles, but we only use the list of predicates to get the biome out of it. I suggest simply bypassing the indirection.

Reply:
I see. Changed it to a `List<Biome>`